atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraReadEteTest.java

```diff
@@ -0,0 +1,147 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.ete;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.palantir.atlasdb.keyvalue.api.RangeRequest;
+import com.palantir.atlasdb.keyvalue.api.RowResult;
+import com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException;
+import com.palantir.atlasdb.table.description.ValueType;
+import com.palantir.atlasdb.todo.ImmutableTodo;
+import com.palantir.atlasdb.todo.Todo;
+import com.palantir.atlasdb.todo.TodoSchema;
+import com.palantir.common.base.BatchingVisitable;
+
+public class QosCassandraReadEteTest extends QosCassandraEteTestSetup {
+    private static final int ONE_TODO_SIZE_IN_BYTES = 1050;
+
+    @BeforeClass
+    public static void createTransactionManagerAndWriteData() {
+        ensureTransactionManagerIsCreated();
+        writeNTodosOfSize(200, 1_000);
+        serializableTransactionManager.close();
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesIfDoesNotExceedLimit() {
+        assertThat(readOneBatchOfSize(1)).hasSize(1);
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50).forEach(i -> assertThat(readOneBatchOfSize(1)).hasSize(1));
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitFirstTime() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+        // The second read might actually be faster as the transaction/metadata is cached
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldNotBeAbleToReadLargeAmountsIfSoftLimitSleepWillBeMoreThanConfiguredBackoffTime() {
+        assertThatThrownBy(() -> readOneBatchOfSize(200))
+                .isInstanceOf(RateLimitExceededException.class)
+                .hasMessage("Rate limited. Available capacity has been exhausted.");
+    }
```

Comment:
We could also have a test to establish that we can be rate limited on one read and then have a smaller read succeed.

Reply:
I dont think thats the behavior we will have, but yes we can a test for reads/writes in this order.

Reply:
Done.